lib/messageReceiver.ts

```diff
@@ -163,29 +164,99 @@ export class MessageReceiver extends LinkEntity {
     });
     if (!options) options = {};
     this.receiverType = receiverType;
+    this.receiveMode = options.receiveMode || ReceiveMode.peekLock;
     this.maxConcurrentCalls = options.maxConcurrentCalls != undefined ?
       options.maxConcurrentCalls : 1;
     this.autoComplete = !!options.autoComplete;
-    this.receiveMode = options.receiveMode || ReceiveMode.peekLock;
+    this.maxAutoRenewDurationInSeconds = options.maxAutoRenewDurationInSeconds != undefined
+      ? options.maxAutoRenewDurationInSeconds
+      : 300;
+    this.autoRenewLock = this.maxAutoRenewDurationInSeconds > 0 && this.receiveMode === ReceiveMode.peekLock;
     this._onAmqpMessage = async (context: EventContext) => {
       const connectionId = this._context.namespace.connectionId;
       const bMessage: ServiceBusMessage = new ServiceBusMessage(this._context, context.message!,
         context.delivery!);
+      let timer: NodeJS.Timer | undefined = undefined;
+      let continueExecution: boolean = true;
+      if (this.autoRenewLock) {
+        // - We need to renew locks before they expire by looking at bMessage.lockedUntilUtc.
+        // - This autorenewal needs to happen **NO MORE** than maxAutoRenewDurationInSeconds
+        // - We should be able to clear the renewal timer when the user's message handler
+        // is done (whether it succeeds or fails).
+        log.receiver("[%s] message with id '%s' is locked until %s.",
+          connectionId, bMessage.messageId, bMessage.lockedUntilUtc!.toString());
```

Comment:
Is `lockedUntilUtc` set by Service Bus itself? If yes, then you can run into problems with clock skew between the client and the server. I don't remember what information is available, but i would base your timing on the duration of the lock rather than actual timestamps, particularly if the timestamp is generated by Service Bus. If the timestamp is generated by you this isn't as big of a deal (though still potentially a problem in certain situations).